src/pal/attestation.cpp

```diff
@@ -0,0 +1,455 @@
+    measurement.data =
+      std::vector<uint8_t>(s_measurement.begin(), s_measurement.end());
+    report_data = VirtualAttestationReportData(
+      j["report_data"].get<std::vector<uint8_t>>());
+  }
+
+  std::optional<long> get_integer_from_cert_extensions(
+    const ccf::crypto::OpenSSL::Unique_X509& x509, const std::string& oid)
+  {
+    Unique_ASN1_OBJECT target(OBJ_txt2obj(oid.c_str(), 1), ASN1_OBJECT_free);
+
+    size_t ext_loc = X509_get_ext_by_OBJ(x509, target, -1);
+    if (ext_loc < 0)
+    {
+      LOG_FAIL_FMT("TCB version OID {} not present in VCEK certificate", oid);
+      return std::nullopt;
+    }
+
+    X509_EXTENSION* ext = X509_get_ext(x509, ext_loc);
+    if (ext == nullptr)
+    {
+      throw std::logic_error(fmt::format(
+        "Expected TCB version OID {} present but could not fetch extension "
+        "at index {} in VCEK certificate",
+        oid,
+        ext_loc));
+    }
+
+    ASN1_OCTET_STRING* data = X509_EXTENSION_get_data(ext);
+    if (data == nullptr)
+    {
+      throw std::logic_error((fmt::format(
+        "Expected TCB version OID {} present but no data in VCEK certificate",
+        oid)));
+    }
+    int len = ASN1_STRING_length(data);
+    const unsigned char* p = ASN1_STRING_get0_data(data);
+
+    Unique_ASN1_INTEGER ai(
+      d2i_ASN1_INTEGER(nullptr, &p, len), ASN1_INTEGER_free);
+    return ASN1_INTEGER_get(ai);
+  }
+
+// Macro to factor repeated pattern (OID lookup -> assign -> early return)
+#define RETRIEVE_TCB_FIELD(TCB, FIELD, OID) \
+  do \
+  { \
+    auto val_##FIELD = get_integer_from_cert_extensions(x509, OID); \
+    if (!val_##FIELD.has_value()) \
+    { \
+      return std::nullopt; \
+    } \
+    if (val_##FIELD.value() < 0 || val_##FIELD.value() > UINT8_MAX) \
+    { \
+      throw std::logic_error(fmt::format( \
+        "Invalid {} value in TCB version: {}", #FIELD, val_##FIELD.value())); \
+    } \
+    (TCB)->FIELD = static_cast<uint8_t>(val_##FIELD.value()); \
+  } while (0)
+
+  std::optional<snp::TcbVersionRaw> get_milan_genoa_tcb_from_cert(
+    const crypto::Pem& vcek_leaf_cert)
+  {
+    // From "Versioned Chip Endorsement Key (VCEK) Certificate and KDS Interface
+    // Specification"
+    // https://www.amd.com/content/dam/amd/en/documents/epyc-technical-docs/specifications/57230.pdf
+    //
+    // Table 10 VCEK Certificate Extensions for Family 19h
+    // 1.3.6.1.4.1.3704.1.3.1 blSPL
+    // 1.3.6.1.4.1.3704.1.3.2 teeSPL
+    // 1.3.6.1.4.1.3704.1.3.4 spl_4
+    // 1.3.6.1.4.1.3704.1.3.5 spl_5
+    // 1.3.6.1.4.1.3704.1.3.6 spl_6
+    // 1.3.6.1.4.1.3704.1.3.7 spl_7
+    // 1.3.6.1.4.1.3704.1.3.3 snpSPL
+    // 1.3.6.1.4.1.3704.1.3.8 ucodeSPL
+
+    snp::TcbVersionRaw raw;
+    auto* tcb = raw.as_milan_genoa();
+
+    ccf::crypto::OpenSSL::Unique_BIO mem_bio(vcek_leaf_cert);
+    ccf::crypto::OpenSSL::Unique_X509 x509(mem_bio, true);
+
+    const std::string tcb_oid_prefix = "1.3.6.1.4.1.3704.1.3.";
```

Comment:
Arguably preferable to use defines here to make sure this is all constexpr:

```
#define TCB_OID_PREFIX "1.3.6.1.4.1.3704.1.3."
#define TCB_BOOT_LOADER TCB_OID_PREFIX "1"
etc
```

Eventually there will be a sensible way to do constexpr things to std::strings, but it doesn't look like it's happened yet.